In [1]:
import pandas as pd
data = pd.read_csv(r"C:\Users\Josh\Documents\Data Science\Autos\autos.csv", encoding='cp1252')


#Let's take a look at the kind of fields contained in the dataset
data.head(10)

dateCrawled                                               name  \
0  2016-03-24 11:52:17                                         Golf_3_1.6   
1  2016-03-24 10:58:45                               A5_Sportback_2.7_Tdi   
2  2016-03-14 12:52:21                     Jeep_Grand_Cherokee_"Overland"   
3  2016-03-17 16:54:04                                 GOLF_4_1_4__3TÜRER   
4  2016-03-31 17:25:20                     Skoda_Fabia_1.4_TDI_PD_Classic   
5  2016-04-04 17:36:23  BMW_316i___e36_Limousine___Bastlerfahrzeug__Ex...   
6  2016-04-01 20:48:51                        Peugeot_206_CC_110_Platinum   
7  2016-03-21 18:54:38                       VW_Derby_Bj_80__Scheunenfund   
8  2016-04-04 23:42:13               Ford_C___Max_Titanium_1_0_L_EcoBoost   
9  2016-03-17 10:53:50  VW_Golf_4_5_tuerig_zu_verkaufen_mit_Anhaengerk...   

   seller offerType  price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot    480     test         NaN                1993   
1  privat   Angebot  18300     test       coupe                2011   
2  privat   Angebot   9800     test         suv                2004   
3  privat   Angebot   1500     test  kleinwagen                2001   
4  privat   Angebot   3600     test  kleinwagen                2008   
5  privat   Angebot    650     test   limousine                1995   
6  privat   Angebot   2200     test      cabrio                2004   
7  privat   Angebot      0     test   limousine                1980   
8  privat   Angebot  14500  control         bus                2014   
9  privat   Angebot    999     test  kleinwagen                1998   

     gearbox  powerPS    model  kilometer  monthOfRegistration fuelType  \
0    manuell        0     golf     150000                    0   benzin   
1    manuell      190      NaN     125000                    5   diesel   
2  automatik      163    grand     125000                    8   diesel   
3    manuell       75     golf     150000                    6   benzin   
4    manuell       69    fabia      90000                    7   diesel   
5    manuell      102      3er     150000                   10   benzin   
6    manuell      109  2_reihe     150000                    8   benzin   
7    manuell       50   andere      40000                    7   benzin   
8    manuell      125    c_max      30000                    8   benzin   
9    manuell      101     golf     150000                    0      NaN   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0  volkswagen               NaN  2016-03-24 00:00:00             0   
1        audi                ja  2016-03-24 00:00:00             0   
2        jeep               NaN  2016-03-14 00:00:00             0   
3  volkswagen              nein  2016-03-17 00:00:00             0   
4       skoda              nein  2016-03-31 00:00:00             0   
5         bmw                ja  2016-04-04 00:00:00             0   
6     peugeot              nein  2016-04-01 00:00:00             0   
7  volkswagen              nein  2016-03-21 00:00:00             0   
8        ford               NaN  2016-04-04 00:00:00             0   
9  volkswagen               NaN  2016-03-17 00:00:00             0   

   postalCode             lastSeen  
0       70435  2016-04-07 03:16:57  
1       66954  2016-04-07 01:46:50  
2       90480  2016-04-05 12:47:46  
3       91074  2016-03-17 17:40:17  
4       60437  2016-04-06 10:17:21  
5       33775  2016-04-06 19:17:07  
6       67112  2016-04-05 18:18:39  
7       19348  2016-03-25 16:47:58  
8       94505  2016-04-04 23:42:13  
9       27472  2016-03-31 17:17:06

In [2]:
#and also a description of the numerical fields
data.describe()

price  yearOfRegistration        powerPS      kilometer  \
count  3.715280e+05       371528.000000  371528.000000  371528.000000   
mean   1.729514e+04         2004.577997     115.549477  125618.688228   
std    3.587954e+06           92.866598     192.139578   40112.337051   
min    0.000000e+00         1000.000000       0.000000    5000.000000   
25%    1.150000e+03         1999.000000      70.000000  125000.000000   
50%    2.950000e+03         2003.000000     105.000000  150000.000000   
75%    7.200000e+03         2008.000000     150.000000  150000.000000   
max    2.147484e+09         9999.000000   20000.000000  150000.000000   

       monthOfRegistration  nrOfPictures    postalCode  
count        371528.000000      371528.0  371528.00000  
mean              5.734445           0.0   50820.66764  
std               3.712412           0.0   25799.08247  
min               0.000000           0.0    1067.00000  
25%               3.000000           0.0   30459.00000  
50%               6.000000           0.0   49610.00000  
75%               9.000000           0.0   71546.00000  
max              12.000000           0.0   99998.00000

In [3]:
#Now to drop columns of the data set that, based on intuition, have little to no predictive power

data.drop(['name','dateCrawled','seller', 'offerType', 'abtest', 'dateCreated', 'nrOfPictures', 'lastSeen','postalCode'], axis = 'columns', inplace=True)

#dateCrawled isn't relevant to sales price; seller is entirely or almost entirely one value, likewise with offerType; 
#I don't understand how abtest is being used here and is probably not helpful; dateCreated is also not relevant to the sales price; 
#nr of pictures is almost entirely 0;lastSeen is also irrelevant; postal code is irrelevant; name is redundant when we have make, model, and brand

In [4]:
# Let's drop any duplicate entries to make the data set more lean
data.drop_duplicates()
#Next, I restrict the domain of the factors to reasonably reduce the effect of outliers and impossible data points that may have been inputted erroneously
data = data[(data.price>= 1000) & (data.price<=200000) & (data.yearOfRegistration>=1950) & (data.yearOfRegistration<=2017) & (data.powerPS>0)]

data.isnull().sum()

price                      0
vehicleType            11916
yearOfRegistration         0
gearbox                 4012
powerPS                    0
model                   9083
kilometer                  0
monthOfRegistration        0
fuelType               12021
brand                      0
notRepairedDamage      31765
dtype: int64

In [5]:
#There's a lot of null values in some of these fields, let's think about how to handle them...
#All of the fields with nulls are categorical, which makes attempting to impute values a bad idea.
#I'll simply assign a category of "null" to these.  Because there's a relatively low amount of nulls compared to the amount of data

data['vehicleType'].fillna(value='null', inplace=True)
data['gearbox'].fillna(value='null', inplace=True)
data['model'].fillna(value='null', inplace=True)
data['fuelType'].fillna(value='null', inplace=True)
data['notRepairedDamage'].fillna(value='null', inplace=True)
data.isnull().sum()

price                  0
vehicleType            0
yearOfRegistration     0
gearbox                0
powerPS                0
model                  0
kilometer              0
monthOfRegistration    0
fuelType               0
brand                  0
notRepairedDamage      0
dtype: int64

In [6]:
features = ['price','yearOfRegistration','powerPS','kilometer','monthOfRegistration','vehicleType','gearbox','model','fuelType','brand','notRepairedDamage']
encoded_data = pd.get_dummies(data[features])
encoded_data.corr()

#One hot encoding the features gives a decent idea of the predictive power of the individual categories.
#There's a ton of dummy variables, but they seem to be generally correlated with price and so I'll use them in my model

price  yearOfRegistration   powerPS  kilometer  \
price                   1.000000            0.293099  0.176594  -0.420562   
yearOfRegistration      0.293099            1.000000  0.029232  -0.311600   
powerPS                 0.176594            0.029232  1.000000   0.004735   
kilometer              -0.420562           -0.311600  0.004735   1.000000   
monthOfRegistration     0.020493            0.019348  0.004094  -0.007837   
vehicleType_andere     -0.018221           -0.046269 -0.005916  -0.001427   
vehicleType_bus        -0.000674            0.023520 -0.024107   0.041624   
vehicleType_cabrio      0.121625           -0.091111  0.027086  -0.080409   
vehicleType_coupe       0.150109           -0.102056  0.071026  -0.014702   
vehicleType_kleinwagen -0.180865            0.028454 -0.137064  -0.156825   
vehicleType_kombi      -0.025293            0.032035  0.034327   0.142945   
vehicleType_limousine  -0.018350           -0.108436  0.033316   0.045092   
vehicleType_null       -0.068493            0.280646 -0.011161   0.042206   
vehicleType_suv         0.166956            0.054638  0.047277  -0.058799   
gearbox_automatik       0.277957            0.053750  0.146489  -0.018758   
gearbox_manuell        -0.264490           -0.051147 -0.140529   0.018823   
gearbox_null           -0.029441           -0.005688 -0.011414  -0.001526   
model_100              -0.016041           -0.061765  0.000183   0.013751   
model_145              -0.004103           -0.005431 -0.000658   0.001754   
model_147              -0.021179            0.002790  0.003907   0.010810   
model_156              -0.020822           -0.011808  0.002681   0.019960   
model_159              -0.000298            0.012245  0.004921   0.004499   
model_1_reihe          -0.022095            0.007477 -0.015346  -0.024675   
model_1er               0.054797            0.072988  0.009287  -0.043300   
model_200              -0.002799           -0.018836  0.002096   0.003500   
model_2_reihe          -0.050042            0.015297 -0.025001  -0.016607   
model_300c              0.008652            0.006911  0.012057   0.006894   
model_3_reihe          -0.030593            0.009303 -0.008739   0.001000   
model_3er              -0.014744           -0.068613  0.032693   0.094372   
model_4_reihe          -0.014632           -0.003532  0.001243   0.013126   
...                          ...                 ...       ...        ...   
brand_jaguar            0.034812           -0.026581  0.024140   0.001377   
brand_jeep              0.024870           -0.017837  0.012614  -0.002511   
brand_kia              -0.005574            0.044443 -0.004270  -0.038746   
brand_lada             -0.009116           -0.004772 -0.006748  -0.030257   
brand_lancia           -0.006250           -0.013847 -0.004164  -0.005729   
brand_land_rover        0.060563           -0.000044  0.006862  -0.004478   
brand_mazda            -0.028854           -0.000035 -0.007577  -0.005666   
brand_mercedes_benz     0.077533           -0.081186  0.062233   0.063121   
brand_mini              0.033712            0.051673 -0.005190  -0.075853   
brand_mitsubishi       -0.022015           -0.002174 -0.002102  -0.007382   
brand_nissan           -0.010975            0.018463 -0.010297  -0.029388   
brand_opel             -0.109621           -0.000101 -0.038756   0.003589   
brand_peugeot          -0.063442            0.023408 -0.026394  -0.013275   
brand_porsche           0.324748           -0.032668  0.070470  -0.044297   
brand_renault          -0.083099            0.015570 -0.036267  -0.009761   
brand_rover            -0.013775           -0.015429 -0.002162   0.005154   
brand_saab             -0.010121           -0.022661  0.005634   0.016195   
brand_seat             -0.024862            0.045881 -0.020663  -0.026547   
brand_skoda            -0.005071            0.062892 -0.020219  -0.035564   
brand_smart            -0.050297            0.030137 -0.044833  -0.081536   
brand_sonstige_autos   

In [7]:
#Now to make the model.  

from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_absolute_error
from sklearn.model_selection import train_test_split


x = encoded_data.drop(['price'], axis = 'columns', inplace= False)
y = encoded_data['price']

train_x, val_x, train_y, val_y = train_test_split(x,y, random_state=1)
model = RandomForestRegressor(random_state=1)
model.fit(train_x,train_y)
val_predictions = model.predict(val_x)

C:\Users\Josh\AppData\Roaming\Python\Python36\site-packages\sklearn\ensemble\forest.py:248: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


In [8]:
print ("Score: %f" % model.score(val_x,val_y))

Score: 0.834912


So, I was able to create a model with 83.5% predictive power on the testing data.  This was a rudimentary exercise to simply get experience parsing through data and creating a usable model with it.  

Some questions/ thoughts for future projects:
    How can I conjugate the dummy variables back into one for use in data visualization?
    How can I manually label and convert numbers into categories for use in predictions (I.E. to make use of the postal code field which likely has predictive power)?
    What other algorithms might be helpful or superior here?
    Spend time learning data visualization to actually show if the process.